src/core/search/search.cc

```diff
@@ -31,69 +31,42 @@ AstExpr ParseQuery(std::string_view query) {
 struct IndexResult {
   using DocVec = vector<DocId>;
 
-  IndexResult() : value{DocVec{}} {};
+  IndexResult() : value_{DocVec{}} {};
 
-  IndexResult(const DocVec* dv) : value{dv} {
+  IndexResult(const DocVec* dv) : value_{dv} {
     if (dv == nullptr)
-      value = DocVec{};
+      value_ = DocVec{};
   }
 
-  IndexResult(DocVec&& dv) : value{move(dv)} {
+  IndexResult(DocVec&& dv) : value_{move(dv)} {
   }
 
   // Transparent const access to underlying value
   const DocVec* operator->() const {
-    return holds_alternative<DocVec>(value) ? &get<DocVec>(value) : get<const DocVec*>(value);
+    return holds_alternative<DocVec>(value_) ? &get<DocVec>(value_) : get<const DocVec*>(value_);
   }
 
-  // Move out of owned or copy borrowed
-  DocVec Take() {
-    if (holds_alternative<DocVec>(value))
-      return move(get<DocVec>(value));
-    return *get<const DocVec*>(value);
-  }
-
-  // Move out of owned or return empty
-  DocVec TakeOwnedOrEmpty() {
-    if (holds_alternative<DocVec>(value))
-      return move(get<DocVec>(value));
-    return DocVec{};
+  IndexResult& operator=(DocVec&& entries) {
+    if (holds_alternative<DocVec>(value_)) {
+      swap(get<DocVec>(value_), entries);  // swap to keep backing array
```

Comment:
Sorry, I'm missing something, what's the advantage of `swap` here? (i.e. which backing array is kept like this which is not otherwise)?

Reply:
Do you mean that of `entries`? Is it important for reusability?

Reply:
Yes, its basically the following

```
current |
          --- [merge to] ---> tmp ---- [move to] ---> current
matched |
```

In the last step, I drop current's backing array if it has one. I can instead store it inside tmp, so it can be re-used. It'll be important for large queries where those arrays can contain thousands of ids